Gracefully handle invalid/corrupt data in the persistent cache

mariospr opened the pull request:
This patch set improves the way we deal with situations where invalid/corrupt data makes it to the persistent cache file in three ways:
  * Discard invalid data found after a valid chunk when reading from the cache, and just report that.
  * Purge the whole persistent cache file in cases where the corruption can't be detected earlier.
  * Send metrics events to the server so that we can track down when these situations happen

Note that is still unclear to me how we actually ended up in this scenario, which is still the big question here, but these changes should at least prevent the daemon from crashing and make it able to keep recording and sending metrics after having detected and deal with these kind of situations.

https://phabricator.endlessm.com/T12739

ptomato commented on daemon/emer-circular-file.c:
```diff
@@ -734,6 +734,19 @@ emer_circular_file_read (EmerCircularFile *self,
+      /* Reading a zero-sized element here means that we have invalid
+       * data ahead, in which case we need to update the priv->size
+       * pointer so that the next time this is run it does not include
+       * the region of invalid data existing after this poing.
```
typo: poing -> point

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -574,7 +574,22 @@ add_stored_events_to_builders (EmerDaemon        *self,
   if (!read_succeeded)
     {
-      g_warning ("Could not read from persistent cache: %s.", error->message);
+      if (error->domain == G_IO_ERROR && error->code == G_IO_ERROR_INVALID_DATA)
```
Use `g_error_matches()`

mariospr replied:
Ok

cosimoc commented on daemon/emer-persistent-cache.c:
```diff
@@ -1051,29 +1051,47 @@ emer_persistent_cache_read (EmerPersistentCache *self,
-  GVariant **local_variants = g_new (GVariant *, num_elems);
-  for (gsize i = 0; i < num_elems; i++)
+  corrupt_data = FALSE;
+  local_variants = g_new (GVariant *, num_elems);
```
Declare these two variables here (I know the coding style here is a bit different from the usual C, but we should keep it consistent)

mariospr replied:
Oh! My mistake. I was under the impression that it was using the C89 way of declaring everything at the beginning of the block. will change it

cosimoc commented on daemon/emer-persistent-cache.c:
```diff
@@ -1086,6 +1104,28 @@ emer_persistent_cache_read (EmerPersistentCache *self,
+
+      *variants = NULL;
+      *num_variants = 0;
+      *token = 0;
```
There's no reason to set these when returning `FALSE`

mariospr replied:
You'r right, I'll remove it

cosimoc commented on daemon/emer-persistent-cache.c:
```diff
@@ -1086,6 +1104,28 @@ emer_persistent_cache_read (EmerPersistentCache *self,
+      gsize j = 0;
+
+      while (j < num_elems)
+        g_bytes_unref (elems[j++]);
```
The bytes in `elems` with index between `0` and `i - 1` have already been unreferenced in the for loop above before encountering the corrupt data. So I think that `j` should be initialized to the value of `i` before the cycle, not `0`.

mariospr replied:
That was actually my intention, will fix

cosimoc commented on daemon/emer-persistent-cache.c:
```diff
@@ -1086,6 +1104,28 @@ emer_persistent_cache_read (EmerPersistentCache *self,
+      g_free (elems);
+
+      while (j < i)
+        g_variant_unref (local_variants[j++]);
```
I think `j` should be initialized to `0` before this loop, since the variants to free in `local_variants` have index between `0` and `i - 1`

mariospr replied:
another mistake, will fix

mariospr commented:
@cosimoc I addressed your comments and this is now ready for re-review. In the meantime I'll work in the alternative version of it that also sends a metric event when this problem is found, but I'll keep this around in case it's important to merge something today and I can't have the other one ready yet.

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -548,6 +566,89 @@ add_events_to_builders (GVariant       **events,
+  /* Do the report in a new iteration of the main loop to make sure we don't
+   * report the event before having finished processing the current cache.
+   */
+  g_idle_add ((GSourceFunc) report_invalid_data_in_cache_on_idle, callback_data);
```
Should probably save the source ID and cancel the timeout in case the daemon is destroyed.

mariospr replied:
Sure, will do.

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -596,6 +699,18 @@ add_stored_events_to_builders (EmerDaemon        *self,
+  if (has_invalid)
+    {
+      GVariant *payload = g_variant_new ("(tt)", num_variants, token);
+      g_variant_ref_sink (payload);
```
I don't think this is needed... `emer_daemon_record_singular_event()` will consume the floating ref

mariospr replied:
Are you sure about that? I can see get_nullable_payload() consuming the floating ref only for the case where `has_payload == FALSE` (meaning that it sinks the ref of the "empty auxiliary payload" is it calls it), which won't be the case here as we will be passing a specific payload whose ref needs sinking at some point.

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -548,6 +566,89 @@ add_events_to_builders (GVariant       **events,
+  CacheMetricEventData *callback_data = g_new (CacheMetricEventData, 1);
+  callback_data->daemon = self;
+  callback_data->event_id = event_id;
+  callback_data->payload = payload != NULL ? g_variant_ref (payload) : NULL;
```
I don't think you need to ref this, unless you want to use the variant after the idle.

mariospr replied:
In general I like to ref objects when sending them as data for asynchronous callbacks, but it's true in this particular case the idiom is not strictly needed since no one will be unreffing

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -548,6 +566,89 @@ add_events_to_builders (GVariant       **events,
+    {
+      GVariant *unboxed_variant = g_variant_new_boolean (FALSE);
+      GVariant *empty_payload = g_variant_new_variant (unboxed_variant);
+      actual_payload = g_variant_ref_sink (empty_payload);
```
I don't think this is needed.

mariospr replied:
I'm not completely sure either, I borrowed this [idiom from emtr_event_recorder_init](https://github.com/endlessm/eos-metrics/blob/27e4194ed5ca1c5966261ef1fed5fa182a4285a8/eosmetrics/emtr-event-recorder.c#L172), as that's how it constructs the empty payload for some reason, and assumed that was the right thing to do as I can't see the server side of things. But I'm happy to write it simpler if possible

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -548,6 +566,89 @@ add_events_to_builders (GVariant       **events,
+                                     actual_payload);
+ free:
+  if (actual_payload != NULL)
+    g_variant_unref (actual_payload);
```
There should be no need to unref this if you follow my other suggestions

mariospr replied:
I'm not sure really. I think it's safer to call g_variant_ref_sink() only (i.e. no other calls to g_variant_ref in other places) for each of the two cases [in the if above this one](https://github.com/endlessm/eos-event-recorder-daemon/pull/214/commits/4a5c09295d8875669b3bd98d7f8025348f184228#diff-3caff38bd400df071bb8b3e5d6ed8d96R611) and then unref here.

cosimoc replied:
I'm pretty sure that `emer_daemon_record_singular_event()` consumes floating refs like e.g. gdbus methods do. See e.g. https://github.com/endlessm/eos-desktop/blob/master/src/shell-app-system.c#L530

mariospr replied:
Ok. I'll believe you :)

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -1115,6 +1248,7 @@ emer_daemon_finalize (GObject *object)
   EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
 
   g_source_remove (priv->upload_events_timeout_source_id);
+  g_source_remove (priv->report_invalid_cache_data_source_id);
```
You have to check if this is not zero before calling this I think.

mariospr replied:
ok

cosimoc commented on daemon/emer-daemon.c:
```diff
@@ -548,6 +567,92 @@ add_events_to_builders (GVariant       **events,
+
+static gboolean
+report_invalid_data_in_cache_on_idle (CacheMetricEventData *callback_data)
+{
```
Need to set the source ID to zero in this callback
